### PR TITLE
Enable Resume Training for Object Detection

### PR DESCRIPTION
**Pull request by Joeper214**
## Motivation and Context

Currently, Models for Object detection Task us using Memory when saving the seen batches during training. The goal of this PR is to save the steps using checkpoint by using `global_steps` from TensorFlow variable.

## Description

- Update object detection models configs to use LOSS_WARMUP_STEPS
- Use global step in YoloV2 loss function
- Update blueoil config templates

## How has this been tested?

This should work by running normal object detection tasks.

for example:

- In blueoil
`$ ./blueoil.sh train config/object_detection.yml`

- In Lmnet
`$ python lmnet/executor/train.py -c lmnet/configs/example/object_detection.py`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Joeper214** [comment]
run blueoil test

File: lmnet/lmnet/networks/object_detection/yolo_v2.py [hunk 1]
```diff
@@ -1194,12 +1194,12 @@ def _one_iou(self, box1, box2):
...
 
     def __calculate_truth_and_maskes(
-            self, gt_boxes_list, predict_boxes, num_cell, image_size, is_training,
+            self, gt_boxes_list, predict_boxes, num_cell, image_size, global_step,
             predict_classes=None, predict_confidence=None
     ):
         """Calculate truth and maskes for loss function from gt boxes and predict boxes.
 
-        1. When trained images is less than seen_threshold, set cell_gt_boxes and coordinate_maskes
+        1. When global steps is less than seen_threshold, set cell_gt_boxes and coordinate_maskes
```
Comment (ruimashita) [inline lmnet/lmnet/networks/object_detection/yolo_v2.py]: When global steps is less than `warmup_steps`

File: lmnet/lmnet/networks/object_detection/yolo_v2.py [hunk 2]
```diff
@@ -1248,9 +1245,10 @@ def __calculate_truth_and_maskes(
 
         # extra coordinate loss for early training steps to encourage predictions to match anchor.
         # https://github.com/pjreddie/darknet/blob/2f212a47425b2e1002c7c8a20e139fe0da7489b5/src/region_layer.c#L248
-        if self.seen < self.seen_threshold:
+        if global_step < self.warmup_steps:
             if self.is_debug:
-                print("current seen: {}. To calculate coordinate loss for early training step.".format(self.seen))
+                print("current seen: {}. warmup steps: {}. \
+                To calculate coordinate loss for early training step.".format(self.seen, self.warmup_steps))
```
Comment (ruimashita) [inline lmnet/lmnet/networks/object_detection/yolo_v2.py]: Did you check this line ? `self.seen` can't use.

File: lmnet/configs/example/object_detection.yaml [hunk 3]
```diff
@@ -103,6 +103,7 @@ NETWORK:
     - 1.0e-06
     - 1.0e-07
   LOSS_IOU_THRESHOLD: 0.6
+  LOSS_WARMUP_STEPS: 160.0
```
Comment (ruimashita) [inline lmnet/configs/example/object_detection.yaml]: not float

File: lmnet/lmnet/networks/object_detection/yolo_v2.py [hunk 4]
```diff
@@ -48,7 +48,7 @@ def __init__(
             nms_iou_threshold=0.5,
             nms_max_output_size=100,
             nms_per_class=True,
-            seen_threshold=12800,
+            loss_warmup_steps=100,
```
Comment (ruimashita) [inline lmnet/lmnet/networks/object_detection/yolo_v2.py]: Why do you choose the `100` default value ? Is it from yolov2 paper?

Comment (ruimashita) [reply lmnet/lmnet/networks/object_detection/yolo_v2.py]: @Joeper214 
darknet implementation has `12800` as threshold, 
https://github.com/pjreddie/darknet/blob/f6d861736038da22c9eb0739dca84003c5a5e275/src/region_layer.c#L254
and batch size is `64` when train VOC datasets.
https://github.com/pjreddie/darknet/blob/master/cfg/yolov2-voc.cfg#L6

I recommend the default value of `loss_warmup_steps` is `12800/64 = 200.

**Comment by Joeper214** [comment]
@ruimashita Thanks for the comments. I updated the branch and tested a training with the given config. Please check again. 

This is so far the results I got.

Screen Shot 2019-07-10 at 9 51 03

**Comment by ruimashita** [comment]
@Joeper214 Thanks 👍

File: blueoil/templates/lmnet/object_detection.tpl.py [hunk 5]
```diff
@@ -107,7 +107,7 @@
 NETWORK.SCORE_THRESHOLD = score_threshold
 NETWORK.NMS_IOU_THRESHOLD = nms_iou_threshold
 NETWORK.NMS_MAX_OUTPUT_SIZE = nms_max_output_size
-NETWORK.SEEN_THRESHOLD = 8000
+NETWORK.LOSS_WARMUP_STEPS = 8000 / BATCH_SIZE
```
Comment (ruimashita) [inline blueoil/templates/lmnet/object_detection.tpl.py]: Should we use `int(8000/BATCH_SIZE)`?

Comment (Joeper214) [reply blueoil/templates/lmnet/object_detection.tpl.py]: yeah. I missed this one. I'll change that Thanks!

**Comment by Joeper214** [comment]
@ruimashita I updated the PR again. Thanks
